File: plugins/devkit/devkit-java-tests/testData/inspections/incorrectCeHandling/IncorrectPceHandlingWhenPceCaughtImplicitlyTests.java
import com.intellij.openapi.diagnostic.Logger;
import com.intellij.openapi.progress.ProcessCanceledException;
import com.example.SubclassOfProcessCanceledException;

class IncorrectPceHandlingWhenPceCaughtImplicitlyTests {
  private static final Logger LOG = Logger.getInstance(IncorrectPceHandlingWhenPceCaughtImplicitlyTests.class);

  // tests for ProcessCanceledException

  void throwPce() throws ProcessCanceledException {
    // anything
  }

  void testPceSwallowed() {
    try {
      throwPce();
    } catch (Exception <error descr="'com.intellij.openapi.progress.ProcessCanceledException' must be rethrown. It is thrown by 'throwPce()'.">e</error>) {
      // exception swallowed
    }
  }

  void testPceLogged() {
    try {
      throwPce();
    } catch (Exception e) {
      <error descr="'com.intellij.openapi.progress.ProcessCanceledException' must not be logged. It is thrown by 'throwPce()'.">LOG.error(e)</error>;
      throw e;
    }
  }

  void testSwallowedAndLoggedWithMessageOnInfoLevel() {
    try {
      throwPce();
    } catch (Exception <error descr="'com.intellij.openapi.progress.ProcessCanceledException' must be rethrown. It is thrown by 'throwPce()'.">e</error>) {
      <error descr="'com.intellij.openapi.progress.ProcessCanceledException' must not be logged. It is thrown by 'throwPce()'.">LOG.info("Error occured", e)</error>;
    }
  }

  void testSwallowedAndLoggedOnInfoLevel() {
    try {
      throwPce();
    } catch (Exception <error descr="'com.intellij.openapi.progress.ProcessCanceledException' must be rethrown. It is thrown by 'throwPce()'.">e</error>) {
      <error descr="'com.intellij.openapi.progress.ProcessCanceledException' must not be logged. It is thrown by 'throwPce()'.">LOG.info(e)</error>;
    }
  }

  void testSwallowedAndLoggedOnErrorLevel() {
    try {
      throwPce();
    } catch (Exception <error descr="'com.intellij.openapi.progress.ProcessCanceledException' must be rethrown. It is thrown by 'throwPce()'.">e</error>) {
      <error descr="'com.intellij.openapi.progress.ProcessCanceledException' must not be logged. It is thrown by 'throwPce()'.">LOG.error(e)</error>;
    }
  }

  void testSwallowedAndOnlyExceptionMessageLogged() {
    try {
      throwPce();
    } catch (Exception <error descr="'com.intellij.openapi.progress.ProcessCanceledException' must be rethrown. It is thrown by 'throwPce()'.">e</error>) {
      <error descr="'com.intellij.openapi.progress.ProcessCanceledException' must not be logged. It is thrown by 'throwPce()'.">LOG.error("Error occurred: " + e.getMessage())</error>;
    }
  }

  void testDisjunctionTypesWhenPceIsFirst() {
    try {
      throwPce();
    } catch (RuntimeException | Error <error descr="'com.intellij.openapi.progress.ProcessCanceledException' must be rethrown. It is thrown by 'throwPce()'.">e</error>) {
      <error descr="'com.intellij.openapi.progress.ProcessCanceledException' must not be logged. It is thrown by 'throwPce()'.">LOG.error("Error occurred: " + e.getMessage())</error>;
    }
  }

  void testDisjunctionTypesWhenPceIsSecond() {
    try {
      throwPce();
    } catch (Error | RuntimeException <error descr="'com.intellij.openapi.progress.ProcessCanceledException' must be rethrown. It is thrown by 'throwPce()'.">e</error>) {
      <error descr="'com.intellij.openapi.progress.ProcessCanceledException' must not be logged. It is thrown by 'throwPce()'.">LOG.error("Error occurred: " + e.getMessage())</error>;
    }
  }

  void testPceSwallowedAndMultipleGenericCatchClauses() {
    try {
      throwPce();
    } catch (RuntimeException <error descr="'com.intellij.openapi.progress.ProcessCanceledException' must be rethrown. It is thrown by 'throwPce()'.">e</error>) {
      // exception swallowed
    } catch (Exception e) {
      // exception swallowed
    } catch (Throwable e) {
      // exception swallowed
    }
  }

  void testPceLoggedAndMultipleGenericCatchClauses() {
    try {
      throwPce();
    } catch (RuntimeException e) {
      <error descr="'com.intellij.openapi.progress.ProcessCanceledException' must not be logged. It is thrown by 'throwPce()'.">LOG.error(e)</error>;
      throw e;
    } catch (Exception e) {
      LOG.error(e);
      throw e;
    } catch (Throwable e) {
      LOG.error(e);
      throw e;
    }
  }

  void testNotHandlingOuterTryIfNestedCatchesPce() {
    try {
      // anything
      try {
        throwPce();
      }
      catch (ProcessCanceledException <error descr="'com.intellij.openapi.progress.ProcessCanceledException' must be rethrown">e</error>) {
        <error descr="'com.intellij.openapi.progress.ProcessCanceledException' must not be logged">LOG.error(e)</error>;
      }
    }
    catch (Throwable e) {
      LOG.error(e);
    }
  }



  // tests for ProcessCanceledException inheritor

  void throwPceInheritor() throws SubclassOfProcessCanceledException {
    // anything
  }

  void testPceInheritorSwallowed() {
    try {
      throwPceInheritor();
    } catch (Exception <error descr="'com.intellij.openapi.progress.ProcessCanceledException' inheritor must be rethrown. It is thrown by 'throwPceInheritor()'.">e</error>) {
      // exception swallowed
    }
  }

  void testPceInheritorLogged() {
    try {
      throwPceInheritor();
    } catch (Exception e) {
      <error descr="'com.intellij.openapi.progress.ProcessCanceledException' inheritor must not be logged. It is thrown by 'throwPceInheritor()'.">LOG.error(e)</error>;
      throw e;
    }
  }

  void testPceInheritorSwallowedAndLoggedWithMessageOnInfoLevel() {
    try {
      throwPceInheritor();
    } catch (Exception <error descr="'com.intellij.openapi.progress.ProcessCanceledException' inheritor must be rethrown. It is thrown by 'throwPceInheritor()'.">e</error>) {
      <error descr="'com.intellij.openapi.progress.ProcessCanceledException' inheritor must not be logged. It is thrown by 'throwPceInheritor()'.">LOG.info("Error occured", e)</error>;
    }
  }

  void testPceInheritorSwallowedAndLoggedOnInfoLevel() {
    try {
      throwPceInheritor();
    } catch (Exception <error descr="'com.intellij.openapi.progress.ProcessCanceledException' inheritor must be rethrown. It is thrown by 'throwPceInheritor()'.">e</error>) {
      <error descr="'com.intellij.openapi.progress.ProcessCanceledException' inheritor must not be logged. It is thrown by 'throwPceInheritor()'.">LOG.info(e)</error>;
    }
  }

  void testPceInheritorSwallowedAndLoggedOnErrorLevel() {
    try {
      throwPceInheritor();
    } catch (Exception <error descr="'com.intellij.openapi.progress.ProcessCanceledException' inheritor must be rethrown. It is thrown by 'throwPceInheritor()'.">e</error>) {
      <error descr="'com.intellij.openapi.progress.ProcessCanceledException' inheritor must not be logged. It is thrown by 'throwPceInheritor()'.">LOG.error(e)</error>;
    }
  }

  void testPceInheritorSwallowedAndOnlyExceptionMessageLogged() {
    try {
      throwPceInheritor();
    } catch (Exception <error descr="'com.intellij.openapi.progress.ProcessCanceledException' inheritor must be rethrown. It is thrown by 'throwPceInheritor()'.">e</error>) {
      <error descr="'com.intellij.openapi.progress.ProcessCanceledException' inheritor must not be logged. It is thrown by 'throwPceInheritor()'.">LOG.error("Error occurred: " + e.getMessage())</error>;
    }
  }

  void testDisjunctionTypesWhenPceInheritorIsFirst() {
    try {
      throwPceInheritor();
    } catch (RuntimeException | Error <error descr="'com.intellij.openapi.progress.ProcessCanceledException' inheritor must be rethrown. It is thrown by 'throwPceInheritor()'.">e</error>) {
      <error descr="'com.intellij.openapi.progress.ProcessCanceledException' inheritor must not be logged. It is thrown by 'throwPceInheritor()'.">LOG.error("Error occurred: " + e.getMessage())</error>;
    }
  }

  void testDisjunctionTypesWhenPceInheritorIsSecond() {
    try {
      throwPceInheritor();
    } catch (Error | RuntimeException <error descr="'com.intellij.openapi.progress.ProcessCanceledException' inheritor must be rethrown. It is thrown by 'throwPceInheritor()'.">e</error>) {
      <error descr="'com.intellij.openapi.progress.ProcessCanceledException' inheritor must not be logged. It is thrown by 'throwPceInheritor()'.">LOG.error("Error occurred: " + e.getMessage())</error>;
    }
  }

  void testPceInheritorSwallowedAndMultipleGenericCatchClauses() {
    try {
      throwPceInheritor();
    } catch (RuntimeException <error descr="'com.intellij.openapi.progress.ProcessCanceledException' inheritor must be rethrown. It is thrown by 'throwPceInheritor()'.">e</error>) {
      // exception swallowed
    } catch (Exception e) {
      // exception swallowed
    } catch (Throwable e) {
      // exception swallowed
    }
  }

  void testPceInheritorLoggedAndMultipleGenericCatchClauses() {
    try {
      throwPceInheritor();
    } catch (RuntimeException e) {
      <error descr="'com.intellij.openapi.progress.ProcessCanceledException' inheritor must not be logged. It is thrown by 'throwPceInheritor()'.">LOG.error(e)</error>;
      throw e;
    } catch (Exception e) {
      LOG.error(e);
      throw e;
    } catch (Throwable e) {
      LOG.error(e);
      throw e;
    }
  }

  void testNotHandlingOuterTryIfNestedCatchesPceInheritor() {
    try {
      // anything
      try {
        throwPceInheritor();
      }
      catch (SubclassOfProcessCanceledException <error descr="'com.intellij.openapi.progress.ProcessCanceledException' inheritor must be rethrown">e</error>) {
        <error descr="'com.intellij.openapi.progress.ProcessCanceledException' inheritor must not be logged">LOG.error(e)</error>;
      }
    }
    catch (Throwable e) {
      LOG.error(e);
    }
  }

}

File: plugins/devkit/devkit-kotlin-tests/testData/inspections/incorrectCeHandling/IncorrectPceHandlingWhenPceCaughtImplicitlyTests.kt
import com.example.SubclassOfProcessCanceledException
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.progress.ProcessCanceledException

private val LOG = Logger.getInstance("any")

class IncorrectPceHandlingWhenPceCaughtImplicitlyTests {

  // tests for ProcessCanceledException
  @Throws(ProcessCanceledException::class)
  fun throwPce() {
    // anything
  }

  fun testPceSwallowed() {
    try {
      throwPce()
    }
    catch (<error descr="'com.intellij.openapi.progress.ProcessCanceledException' must be rethrown. It is thrown by 'throwPce()'.">e</error>: Exception) {
      // exception swallowed
    }
  }

  fun testPceLogged() {
    try {
      throwPce()
    }
    catch (e: Exception) {
      LOG.<error descr="'com.intellij.openapi.progress.ProcessCanceledException' must not be logged. It is thrown by 'throwPce()'.">error(e)</error>
      throw e
    }
  }

  fun testSwallowedAndLoggedWithMessageOnInfoLevel() {
    try {
      throwPce()
    }
    catch (<error descr="'com.intellij.openapi.progress.ProcessCanceledException' must be rethrown. It is thrown by 'throwPce()'.">e</error>: Exception) {
      LOG.<error descr="'com.intellij.openapi.progress.ProcessCanceledException' must not be logged. It is thrown by 'throwPce()'.">info("Error occured", e)</error>
    }
  }

  fun testSwallowedAndLoggedOnInfoLevel() {
    try {
      throwPce()
    }
    catch (<error descr="'com.intellij.openapi.progress.ProcessCanceledException' must be rethrown. It is thrown by 'throwPce()'.">e</error>: Exception) {
      LOG.<error descr="'com.intellij.openapi.progress.ProcessCanceledException' must not be logged. It is thrown by 'throwPce()'.">info(e)</error>
    }
  }

  fun testSwallowedAndLoggedOnErrorLevel() {
    try {
      throwPce()
    }
    catch (<error descr="'com.intellij.openapi.progress.ProcessCanceledException' must be rethrown. It is thrown by 'throwPce()'.">e</error>: Exception) {
      LOG.<error descr="'com.intellij.openapi.progress.ProcessCanceledException' must not be logged. It is thrown by 'throwPce()'.">error(e)</error>
    }
  }

  fun testSwallowedAndOnlyExceptionMessageLogged() {
    try {
      throwPce()
    }
    catch (<error descr="'com.intellij.openapi.progress.ProcessCanceledException' must be rethrown. It is thrown by 'throwPce()'.">e</error>: Exception) {
      LOG.<error descr="'com.intellij.openapi.progress.ProcessCanceledException' must not be logged. It is thrown by 'throwPce()'.">error("Error occurred: " + e.message)</error>
    }
  }

  fun testPceSwallowedAndMultipleGenericCatchClauses() {
    try {
      throwPce()
    }
    catch (<error descr="'com.intellij.openapi.progress.ProcessCanceledException' must be rethrown. It is thrown by 'throwPce()'.">e</error>: RuntimeException) {
      // exception swallowed
    }
    catch (e: Exception) {
      // exception swallowed
    }
    catch (e: Throwable) {
      // exception swallowed
    }
  }

  fun testPceLoggedAndMultipleGenericCatchClauses() {
    try {
      throwPce()
    }
    catch (e: RuntimeException) {
      LOG.<error descr="'com.intellij.openapi.progress.ProcessCanceledException' must not be logged. It is thrown by 'throwPce()'.">error(e)</error>
      throw e
    }
    catch (e: Exception) {
      LOG.error(e)
      throw e
    }
    catch (e: Throwable) {
      LOG.error(e)
      throw e
    }
  }

  fun testNotHandlingOuterTryIfNestedCatchesPce() {
    try {
      // anything
      try {
        throwPce()
      }
      catch (<error descr="'com.intellij.openapi.progress.ProcessCanceledException' must be rethrown">e</error>: ProcessCanceledException) {
        LOG.<error descr="'com.intellij.openapi.progress.ProcessCanceledException' must not be logged">error(e)</error>
      }
    }
    catch (e: Throwable) {
      LOG.error(e)
    }
  }

  // tests for ProcessCanceledException inheritor
  @Throws(SubclassOfProcessCanceledException::class)
  fun throwPceInheritor() {
    // anything
  }

  fun testPceInheritorSwallowed() {
    try {
      throwPceInheritor()
    }
    catch (<error descr="'com.intellij.openapi.progress.ProcessCanceledException' inheritor must be rethrown. It is thrown by 'throwPceInheritor()'.">e</error>: Exception) {
      // exception swallowed
    }
  }

  fun testPceInheritorLogged() {
    try {
      throwPceInheritor()
    }
    catch (e: Exception) {
      LOG.<error descr="'com.intellij.openapi.progress.ProcessCanceledException' inheritor must not be logged. It is thrown by 'throwPceInheritor()'.">error(e)</error>
      throw e
    }
  }

  fun testPceInheritorSwallowedAndLoggedWithMessageOnInfoLevel() {
    try {
      throwPceInheritor()
    }
    catch (<error descr="'com.intellij.openapi.progress.ProcessCanceledException' inheritor must be rethrown. It is thrown by 'throwPceInheritor()'.">e</error>: Exception) {
      LOG.<error descr="'com.intellij.openapi.progress.ProcessCanceledException' inheritor must not be logged. It is thrown by 'throwPceInheritor()'.">info("Error occured", e)</error>
    }
  }

  fun testPceInheritorSwallowedAndLoggedOnInfoLevel() {
    try {
      throwPceInheritor()
    }
    catch (<error descr="'com.intellij.openapi.progress.ProcessCanceledException' inheritor must be rethrown. It is thrown by 'throwPceInheritor()'.">e</error>: Exception) {
      LOG.<error descr="'com.intellij.openapi.progress.ProcessCanceledException' inheritor must not be logged. It is thrown by 'throwPceInheritor()'.">info(e)</error>
    }
  }

  fun testPceInheritorSwallowedAndLoggedOnErrorLevel() {
    try {
      throwPceInheritor()
    }
    catch (<error descr="'com.intellij.openapi.progress.ProcessCanceledException' inheritor must be rethrown. It is thrown by 'throwPceInheritor()'.">e</error>: Exception) {
      LOG.<error descr="'com.intellij.openapi.progress.ProcessCanceledException' inheritor must not be logged. It is thrown by 'throwPceInheritor()'.">error(e)</error>
    }
  }

  fun testPceInheritorSwallowedAndOnlyExceptionMessageLogged() {
    try {
      throwPceInheritor()
    }
    catch (<error descr="'com.intellij.openapi.progress.ProcessCanceledException' inheritor must be rethrown. It is thrown by 'throwPceInheritor()'.">e</error>: Exception) {
      LOG.<error descr="'com.intellij.openapi.progress.ProcessCanceledException' inheritor must not be logged. It is thrown by 'throwPceInheritor()'.">error("Error occurred: " + e.message)</error>
    }
  }

  fun testPceInheritorSwallowedAndMultipleGenericCatchClauses() {
    try {
      throwPceInheritor()
    }
    catch (<error descr="'com.intellij.openapi.progress.ProcessCanceledException' inheritor must be rethrown. It is thrown by 'throwPceInheritor()'.">e</error>: RuntimeException) {
      // exception swallowed
    }
    catch (e: Exception) {
      // exception swallowed
    }
    catch (e: Throwable) {
      // exception swallowed
    }
  }

  fun testPceInheritorLoggedAndMultipleGenericCatchClauses() {
    try {
      throwPceInheritor()
    }
    catch (e: RuntimeException) {
      LOG.<error descr="'com.intellij.openapi.progress.ProcessCanceledException' inheritor must not be logged. It is thrown by 'throwPceInheritor()'.">error(e)</error>
      throw e
    }
    catch (e: Exception) {
      LOG.error(e)
      throw e
    }
    catch (e: Throwable) {
      LOG.error(e)
      throw e
    }
  }

  fun testNotHandlingOuterTryIfNestedCatchesPceInheritor() {
    try {
      // anything
      try {
        throwPceInheritor()
      }
      catch (<error descr="'com.intellij.openapi.progress.ProcessCanceledException' inheritor must be rethrown">e</error>: SubclassOfProcessCanceledException) {
        LOG.<error descr="'com.intellij.openapi.progress.ProcessCanceledException' inheritor must not be logged">error(e)</error>
      }
    }
    catch (e: Throwable) {
      LOG.error(e)
    }
  }

}
